framework/decode/openxr_replay_consumer_base.cpp

```diff
@@ -802,16 +832,32 @@ void OpenXrReplayConsumerBase::Process_xrPollEvent(const ApiCallInfo&
     XrInstance         in_instance = MapHandle<OpenXrInstanceInfo>(instance, &CommonObjectInfoTable::GetXrInstanceInfo);
     XrEventDataBuffer* capture_event = eventData->GetPointer();
 
-    // We received events that haven't been handled yet already, so see if this one is in the list already
-    for (size_t ii = 0; ii < received_events_.size(); ++ii)
+    // If we have already received some events, check this event against one of the previously
+    // occurring events.  If it has already occurred, remove it from the list and return.
+    for (size_t ii = 0; ii < unhandled_events_from_replay_.size(); ++ii)
     {
-        if (received_events_[ii].type == capture_event->type)
+        if (unhandled_events_from_replay_[ii].type == capture_event->type)
         {
             GFXRECON_LOG_WARNING("Previously received event %s (0x%x, %u)",
                                  GetEventTypeString(capture_event->type),
                                  capture_event->type,
                                  capture_event->type);
-            received_events_.erase(received_events_.begin() + ii);
+            unhandled_events_from_replay_.erase(unhandled_events_from_replay_.begin() + ii);
+            return;
+        }
+    }
+
+    // If we have events that occurred during capture previously, but were not detected during
+    // replay, see if this incoming event is one of those.  If so, remove it from the list and return.
+    for (size_t ii = 0; ii < unhandled_events_from_capture_.size(); ++ii)
+    {
+        if (unhandled_events_from_capture_[ii].type == capture_event->type)
+        {
+            GFXRECON_LOG_WARNING("Previously recorded but skipped event %s (0x%x, %u)",
+                                 GetEventTypeString(capture_event->type),
+                                 capture_event->type,
+                                 capture_event->type);
+            unhandled_events_from_capture_.erase(unhandled_events_from_capture_.begin() + ii);
```

Comment:
same question -- why everything up to the currently found event?

Reply:
Shouldn't the unhandled_event_from_capture be compared with outEventData in the while loop before?

This is just weeding out the current capture event if the previous instance of that event was unhandled.  If we are just going to keep the last event of a given type, that's fine, but then we shouldn't return, and go look for the new one, having simply erased the old

If we're going to attempt to handle all event in order of capture (up to the LIMIT above), then, this check should be in the while loop, and a match should erase the unhandled_event_from_capture, and retry (maybe resetting the retry count and interval).

Reply:
>  same question -- why everything up to the currently found event?

Same point, it is an erase on a singular iterator.

> Shouldn't the unhandled_event_from_capture be compared with outEventData in the while loop before?

Why, if we already received the captured the event from a previous  `xrPollEvent` call?  We don't need to spin wait for something we already got later.

> This is just weeding out the current capture event if the previous instance of that event was unhandled. If we are just going to keep the last event of a given type, that's fine, but then we shouldn't return, and go look for the new one, having simply erased the old

I don't follow.  Why not handle the event outside of the loop if we already received one.  That means while we were looking for something else, we received the event that the captured event wanted to encounter now. 

For example, let's say we got 5 unknown and unhandled events during capture and then a `XR_TYPE_EVENT_DATA_SESSION_STATE_CHANGED` event.  During  replay, if we get the `XR_TYPE_EVENT_DATA_SESSION_STATE_CHANGED`  and then the 5 unknown events, your logic would only store one unknown event and then spin on all the rest because it would only save one.

Reply:
Discussed in video chat.